unexpected assertion failed using global index

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
Here is the log from our internal test schrddl. 
```
2024/10/09 16:48:19 ddl.go:212: [0;31m[fatal] [error] [instance 16] ERROR: Error 8141: assertion failed: key: 7480000000000000ba5f69800000000000000307800029e4698b0200, assertion: Exist, start_ts: 453113916671131764, existing start ts: 0, existing commit ts: 0
/build/ddl/dml_ops.go:52: Error when executing SQL: delete from `b5d472af` where not( IsNull( `b5d472af`.`56b0971e` ) ) or `b5d472af`.`56b0971e` in ( 1444517 ,824880 ) 
```

### 2. What did you expect to see? (Required)
No error
### 3. What did you see instead (Required)
assertion failed
### 4. What is your TiDB version? (Required)
master
<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/ddl/index_merge_tmp.go
+++ b/pkg/ddl/index_merge_tmp.go
@@ -350,7 +350,7 @@ func (w *mergeIndexWorker) fetchTempIndexVals(
 
 			// Extract the operations on the original index and replay them later.
 			for _, elem := range tempIdxVal {
-				if elem.KeyVer == tables.TempIndexKeyTypeMerge || elem.KeyVer == tables.TempIndexKeyTypeDelete {
+				if elem.KeyVer == tablecodec.TempIndexKeyTypeMerge || elem.KeyVer == tablecodec.TempIndexKeyTypeDelete {
 					// For 'm' version kvs, they are double-written.
 					// For 'd' version kvs, they are written in the delete-only state and can be dropped safely.
 					continue

--- a/pkg/ddl/ingest/BUILD.bazel
+++ b/pkg/ddl/ingest/BUILD.bazel
@@ -72,7 +72,7 @@ go_test(
     embed = [":ingest"],
     flaky = True,
     race = "on",
-    shard_count = 21,
+    shard_count = 22,
     deps = [
         "//pkg/config",
         "//pkg/ddl/ingest/testutil",

--- a/pkg/ddl/ingest/integration_test.go
+++ b/pkg/ddl/ingest/integration_test.go
@@ -470,3 +470,30 @@ func TestAddGlobalIndexInIngest(t *testing.T) {
 	require.Equal(t, rsGlobalIndex1.String(), rsTable.String())
 	require.Equal(t, rsGlobalIndex1.String(), rsGlobalIndex2.String())
 }
+
+func TestAddGlobalIndexInIngestWithUpdate(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	defer ingesttestutil.InjectMockBackendMgr(t, store)()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int) partition by hash(a) partitions 5")
+	tk.MustExec("insert into t (a, b) values (1, 1), (2, 2), (3, 3)")
+	var i atomic.Int32
+	i.Store(3)
+	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/ddl/onJobUpdated", func(job *model.Job) {
+		tk2 := testkit.NewTestKit(t, store)
+		tmp := i.Add(1)
+		_, err := tk2.Exec(fmt.Sprintf("insert into test.t values (%d, %d)", tmp, tmp))
+		assert.Nil(t, err)
+
+		_, err = tk2.Exec(fmt.Sprintf("update test.t set b = b + 11, a = b where b = %d", tmp-1))
+		assert.Nil(t, err)
+	})
+	tk.MustExec("alter table t add unique index idx(b) global")
+	rsGlobalIndex := tk.MustQuery("select *,_tidb_rowid from t use index(idx)").Sort()
+	rsTable := tk.MustQuery("select *,_tidb_rowid from t use index()").Sort()
+	require.Equal(t, rsGlobalIndex.String(), rsTable.String())
+}

--- a/pkg/kv/key.go
+++ b/pkg/kv/key.go
@@ -683,9 +683,9 @@ func (m *MemAwareHandleMap[V]) Range(fn func(h Handle, val V) bool) {
 			return
 		}
 	}
-	for _, v := range m.partitionInts {
+	for pid, v := range m.partitionInts {
 		for h, val := range v.M {
-			if !fn(IntHandle(h), val) {
+			if !fn(NewPartitionHandle(pid, IntHandle(h)), val) {
 				return
 			}
 		}

--- a/pkg/table/tables/index.go
+++ b/pkg/table/tables/index.go
@@ -194,7 +194,7 @@ func (c *index) create(sctx table.MutateContext, txn kv.Transaction, indexedValu
 		)
 		if !opt.FromBackFill() {
 			key, tempKey, keyVer = GenTempIdxKeyByState(c.idxInfo, key)
-			if keyVer == TempIndexKeyTypeBackfill || keyVer == TempIndexKeyTypeDelete {
+			if keyVer == tablecodec.TempIndexKeyTypeBackfill || keyVer == tablecodec.TempIndexKeyTypeDelete {
 				key, tempKey = tempKey, nil
 				keyIsTempIdxKey = true
 			}
@@ -409,6 +409,10 @@ func (c *index) Delete(ctx table.MutateContext, txn kv.Transaction, indexedValue
 		}
 
 		tempValElem := tablecodec.TempIndexValueElem{Handle: h, KeyVer: tempKeyVer, Delete: true, Distinct: distinct}
+		if c.idxInfo.Global {
+			tempValElem.Global = true
+			tempValElem.PartitionID = c.phyTblID
+		}
 		if distinct {
 			if len(key) > 0 {
 				okToDelete := true
@@ -479,40 +483,29 @@ func (c *index) GenIndexKVIter(ec errctx.Context, loc *time.Location, indexedVal
 	return table.NewPlainIndexKVGenerator(c, ec, loc, h, handleRestoreData, indexedValue)
 }
 
-const (
-	// TempIndexKeyTypeNone means the key is not a temporary index key.
-	TempIndexKeyTypeNone byte = 0
-	// TempIndexKeyTypeDelete indicates this value is written in the delete-only stage.
-	TempIndexKeyTypeDelete byte = 'd'
-	// TempIndexKeyTypeBackfill indicates this value is written in the backfill stage.
-	TempIndexKeyTypeBackfill byte = 'b'
-	// TempIndexKeyTypeMerge indicates this value is written in the merge stage.
-	TempIndexKeyTypeMerge byte = 'm'
-)
-
 // GenTempIdxKeyByState is used to get the key version and the temporary key.
 // The tempKeyVer means the temp index key/value version.
 func GenTempIdxKeyByState(indexInfo *model.IndexInfo, indexKey kv.Key) (key, tempKey kv.Key, tempKeyVer byte) {
 	if indexInfo.State != model.StatePublic {
 		switch indexInfo.BackfillState {
 		case model.BackfillStateInapplicable:
-			return indexKey, nil, TempIndexKeyTypeNone
+			return indexKey, nil, tablecodec.TempIndexKeyTypeNone
 		case model.BackfillStateRunning:
 			// Write to the temporary index.
 			tablecodec.IndexKey2TempIndexKey(indexKey)
 			if indexInfo.State == model.StateDeleteOnly {
-				return nil, indexKey, TempIndexKeyTypeDelete
+				return nil, indexKey, tablecodec.TempIndexKeyTypeDelete
 			}
-			return nil, indexKey, TempIndexKeyTypeBackfill
+			return nil, indexKey, tablecodec.TempIndexKeyTypeBackfill
 		case model.BackfillStateReadyToMerge, model.BackfillStateMerging:
 			// Double write
 			tmp := make([]byte, len(indexKey))
 			copy(tmp, indexKey)
 			tablecodec.IndexKey2TempIndexKey(tmp)
-			return indexKey, tmp, TempIndexKeyTypeMerge
+			return indexKey, tmp, tablecodec.TempIndexKeyTypeMerge
 		}
 	}
-	return indexKey, nil, TempIndexKeyTypeNone
+	return indexKey, nil, tablecodec.TempIndexKeyTypeNone
 }
 
 func (c *index) Exist(ec errctx.Context, loc *time.Location, txn kv.Transaction, indexedValue []types.Datum, h kv.Handle) (bool, kv.Handle, error) {

--- a/pkg/table/tables/index.go
+++ b/pkg/table/tables/index.go
@@ -411,7 +411,7 @@ func (c *index) Delete(ctx table.MutateContext, txn kv.Transaction, indexedValue
 		tempValElem := tablecodec.TempIndexValueElem{Handle: h, KeyVer: tempKeyVer, Delete: true, Distinct: distinct}
 		if c.idxInfo.Global {
 			tempValElem.Global = true
-			tempValElem.PartitionID = c.phyTblID
+			tempValElem.Handle = kv.NewPartitionHandle(c.phyTblID, h)
 		}
 		if distinct {
 			if len(key) > 0 {

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1316,6 +1316,7 @@ type TempIndexValueElem struct {
 	Delete   bool
 	Distinct bool
 
+	// Global means it's a global Index, for partitioned tables. Only for `distinct` + `deleted` scenario right now.
 	Global bool
 }
 

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1307,7 +1307,7 @@ func (v TempIndexValue) FilterOverwritten() TempIndexValue {
 // A temp index value element is encoded as one of:
 //   - [flag 1 byte][value_length 2 bytes ] [value value_len bytes]   [key_version 1 byte] {distinct normal}
 //   - [flag 1 byte][value value_len bytes]                           [key_version 1 byte] {non-distinct normal}
-//   - [flag 1 byte][handle_length 2 bytes] [handle handle_len bytes] [key_version 1 byte] {distinct deleted}
+//   - [flag 1 byte][handle_length 2 bytes] [handle handle_len bytes] [partitionIdFlag 1 byte] [partitionID 8 bytes] [key_version 1 byte] {distinct deleted}
 //   - [flag 1 byte]                                                  [key_version 1 byte] {non-distinct deleted}
 type TempIndexValueElem struct {
 	Value    []byte
@@ -1316,8 +1316,7 @@ type TempIndexValueElem struct {
 	Delete   bool
 	Distinct bool
 
-	Global      bool
-	PartitionID int64
+	Global bool
 }
 
 const (
@@ -1329,8 +1328,8 @@ const (
 	TempIndexKeyTypeBackfill byte = 'b'
 	// TempIndexKeyTypeMerge indicates this value is written in the merge stage.
 	TempIndexKeyTypeMerge byte = 'm'
-	// TempIndexKeyTypePartitionIdFlag indicates the following value is partition id.
-	TempIndexKeyTypePartitionIdFlag byte = 'p'
+	// TempIndexKeyTypePartitionIDFlag indicates the following value is partition id.
+	TempIndexKeyTypePartitionIDFlag byte = 'p'
 )
 
 // Encode encodes the temp index value.
@@ -1359,8 +1358,8 @@ func (v *TempIndexValueElem) Encode(buf []byte) []byte {
 			buf = append(buf, byte(hLen>>8), byte(hLen))
 			buf = append(buf, hEncoded...)
 			if v.Global {
-				buf = append(buf, TempIndexKeyTypePartitionIdFlag)
-				buf = append(buf, codec.EncodeInt(nil, v.PartitionID)...)
+				buf = append(buf, TempIndexKeyTypePartitionIDFlag)
+				buf = append(buf, codec.EncodeInt(nil, v.Handle.(kv.PartitionHandle).PartitionID)...)
 			}
 			buf = append(buf, v.KeyVer)
 			return buf
@@ -1439,13 +1438,14 @@ func (v *TempIndexValueElem) DecodeOne(b []byte) (remain []byte, err error) {
 			v.Handle, _ = kv.NewCommonHandle(b[:hLen])
 		}
 		b = b[hLen:]
-		if b[0] == TempIndexKeyTypePartitionIdFlag {
+		if b[0] == TempIndexKeyTypePartitionIDFlag {
 			v.Global = true
-			_, v.PartitionID, err = codec.DecodeInt(b[1:9])
+			var pid int64
+			_, pid, err = codec.DecodeInt(b[1:9])
 			if err != nil {
 				return nil, err
 			}
-			v.Handle = kv.NewPartitionHandle(v.PartitionID, v.Handle)
+			v.Handle = kv.NewPartitionHandle(pid, v.Handle)
 			b = b[9:]
 		}
 		v.KeyVer = b[0]

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1316,7 +1316,7 @@ type TempIndexValueElem struct {
 	Delete   bool
 	Distinct bool
 
-	// Global means it's a global Index, for partitioned tables. Only for `distinct` + `deleted` scenario right now.
+	// Global means it's a global Index, for partitioned tables. Currently only used in `distinct` + `deleted` scenarios.
 	Global bool
 }
 

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1348,11 +1348,11 @@ func (v *TempIndexValueElem) Encode(buf []byte) []byte {
 			}
 			// flag + handle length + handle + [partition id] + temp key version
 			if buf == nil {
-				len := hLen + 4
+				l := hLen + 4
 				if v.Global {
-					len += 9
+					l += 9
 				}
-				buf = make([]byte, 0, len)
+				buf = make([]byte, 0, l)
 			}
 			buf = append(buf, byte(TempIndexValueFlagDeleted))
 			buf = append(buf, byte(hLen>>8), byte(hLen))

--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1315,7 +1315,23 @@ type TempIndexValueElem struct {
 	KeyVer   byte
 	Delete   bool
 	Distinct bool
+
+	Global      bool
+	PartitionID int64
 }
+
+const (
+	// TempIndexKeyTypeNone means the key is not a temporary index key.
+	TempIndexKeyTypeNone byte = 0
+	// TempIndexKeyTypeDelete indicates this value is written in the delete-only stage.
+	TempIndexKeyTypeDelete byte = 'd'
+	// TempIndexKeyTypeBackfill indicates this value is written in the backfill stage.
+	TempIndexKeyTypeBackfill byte = 'b'
+	// TempIndexKeyTypeMerge indicates this value is written in the merge stage.
+	TempIndexKeyTypeMerge byte = 'm'
+	// TempIndexKeyTypePartitionIdFlag indicates the following value is partition id.
+	TempIndexKeyTypePartitionIdFlag byte = 'p'
+)
 
 // Encode encodes the temp index value.
 func (v *TempIndexValueElem) Encode(buf []byte) []byte {
@@ -1331,13 +1347,21 @@ func (v *TempIndexValueElem) Encode(buf []byte) []byte {
 				hEncoded = handle.Encoded()
 				hLen = uint16(len(hEncoded))
 			}
-			// flag + handle length + handle + temp key version
+			// flag + handle length + handle + [partition id] + temp key version
 			if buf == nil {
-				buf = make([]byte, 0, hLen+4)
+				len := hLen + 4
+				if v.Global {
+					len += 9
+				}
+				buf = make([]byte, 0, len)
 			}
 			buf = append(buf, byte(TempIndexValueFlagDeleted))
 			buf = append(buf, byte(hLen>>8), byte(hLen))
 			buf = append(buf, hEncoded...)
+			if v.Global {
+				buf = append(buf, TempIndexKeyTypePartitionIdFlag)
+				buf = append(buf, codec.EncodeInt(nil, v.PartitionID)...)
+			}
 			buf = append(buf, v.KeyVer)
 			return buf
 		}
@@ -1415,6 +1439,15 @@ func (v *TempIndexValueElem) DecodeOne(b []byte) (remain []byte, err error) {
 			v.Handle, _ = kv.NewCommonHandle(b[:hLen])
 		}
 		b = b[hLen:]
+		if b[0] == TempIndexKeyTypePartitionIdFlag {
+			v.Global = true
+			_, v.PartitionID, err = codec.DecodeInt(b[1:9])
+			if err != nil {
+				return nil, err
+			}
+			v.Handle = kv.NewPartitionHandle(v.PartitionID, v.Handle)
+			b = b[9:]
+		}
 		v.KeyVer = b[0]
 		b = b[1:]
 		v.Distinct = true

